src/Kriti/Parser/Monad.hs

```diff
@@ -0,0 +1,216 @@
+      }
+
+{-# INLINE setInput #-}
+setInput :: AlexInput -> Parser ()
+setInput input = modify' $ \s -> s {parseInput = input}
+
+{-# INLINE getInput #-}
+getInput :: Parser AlexInput
+getInput = gets parseInput
+
+{-# INLINE getParseColumn #-}
+getParseColumn :: Parser Int
+getParseColumn = gets (col . lexPos . parseInput)
+
+{-# INLINE location #-}
+location :: Parser Span
+location = gets parseSpan
+
+{-# INLINE located #-}
+located :: a -> Parser (Loc a)
+located a = do
+  sp <- location
+  pure $ Loc sp a
+
+-------------------
+--- Start Codes ---
+-------------------
+
+-- | Get the current start code.
+startCode :: Parser Int
+startCode = gets (NE.head . parseStartCodes)
+
+-- | Push a new start code to the stack.
+pushStartCode :: Int -> Parser ()
+pushStartCode code = modify' $ \st ->
+  st {parseStartCodes = code NE.<| (parseStartCodes st)}
+
+-- | Pop a start code off the stack.
+popStartCode :: Parser ()
+popStartCode = modify' $ \st ->
+  st
+    { parseStartCodes =
+        case parseStartCodes st of
+          _ NE.:| [] -> 0 NE.:| []
+          _ NE.:| (x : xs) -> x NE.:| xs
+    }
+
+----------------------
+--- Error Handling ---
+----------------------
+
+data ParseError
+  = EmptyTokenStream Span
+  | UnexpectedToken (Loc Token)
+  | InvalidLexeme AlexSourcePos B.ByteString
+  deriving (Show)
+
+instance E.RenderError ParseError where
+  render (EmptyTokenStream s) =
+    E.RenderedError
+      { _code = E.ParseErrorCode,
+        _message = "ParseError: Empty token stream.",
+        _span = s
+      }
+  render (UnexpectedToken tok) =
+    let tok' = serialize $ unLoc tok
+        span' = locate tok
+     in E.RenderedError
+          { _code = E.ParseErrorCode,
+            _message = "ParseError: Unexpected token '" <> tok' <> "'.",
+            _span = span'
+          }
+  render (InvalidLexeme start inp) =
+    E.RenderedError
+      { _code = E.LexErrorCode,
+        _message = "LexError: Invalid Lexeme: '" <> TE.decodeUtf8 inp <> "'",
+        _span = Span start (overCol (+ (B.length inp)) start)
+      }
+
+parseError :: ParseError -> Parser a
+parseError err = throwError err
+
+--------------
+--- Tokens ---
+--------------
+
+-- | Construct a Token from the matched Text and the current `parseSpan`.
+{-# INLINE token #-}
+token :: (Loc T.Text -> Token) -> B.ByteString -> Parser Token
+token k bs = k <$> located (TE.decodeUtf8 bs)
+
+-- | Construct a `(TokenSymbol (Loc _))` using the current `parseSpan`
+-- to construct the `Loc _`.
+{-# INLINE symbol #-}
+symbol :: Symbol -> B.ByteString -> Parser Token
+symbol sym _ = do
+  sp <- location
+  pure $ TokSymbol $ Loc sp sym
+
+-----------------------
+--- Alex Primitives ---
+-----------------------
+
+data AlexInput = AlexInput
+  { lexPos :: AlexSourcePos,
+    lexPrevChar :: Char,
+    -- | current input bytestring
+    lexBytes :: B.ByteString,
+    -- | remaining bytes in current character
+    lexCharBytes :: [Word8]
+  }
+
+{-# INLINE nextLine #-}
```

Comment:
What's the guidance behind these `INLINE` annotations?

Reply:
I didn't look at the core or anything but its a good idea to make sure all of this code gets inlined.